src/datasources/asset-calibration/components/AssetCalibrationQueryEditor.tsx

```diff
@@ -7,24 +7,20 @@ import { enumToOptions } from '../../../core/utils';
 import _ from 'lodash';
 import { AssetCalibrationQueryBuilder } from './AssetCalibrationQueryBuilder';
 import { Workspace } from 'core/types';
-import { FloatingError, parseErrorMessage } from 'core/errors';
 
 type Props = QueryEditorProps<AssetCalibrationDataSource, AssetCalibrationQuery>;
 
 export function AssetCalibrationQueryEditor({ query, onChange, onRunQuery, datasource }: Props) {
   query = datasource.prepareQuery(query) as AssetCalibrationQuery;
 
+  const areWorkspacesLoaded = datasource.workspacesCache.size > 0;
   const [workspaces, setWorkspaces] = useState<Workspace[]>([]);
-  const [error, setError] = useState<string>('');
 
   useEffect(() => {
-    const getWorkspaces = async () => {
-      const workspaces = await datasource.getWorkspaces();
-      setWorkspaces(workspaces);
+    if (areWorkspacesLoaded) {
+      setWorkspaces(Array.from(datasource.workspacesCache.values()));
     }
-
-    getWorkspaces().catch(error => setError(parseErrorMessage(error) || 'Failed to fetch workspaces'));
-  }, [datasource]);
+  }, [datasource, areWorkspacesLoaded]);
```

Comment:
I haven't tested locally but wouldn't datasource.workspacesCache be enough/better as parameters in the dependency array of the useEffect ? This way if the datasource makes another query to the workspaces, you'd get them updated here as well rather than just checking if there are any elements in the list.

Reply:
Changed the approach here due to recent clarifications.